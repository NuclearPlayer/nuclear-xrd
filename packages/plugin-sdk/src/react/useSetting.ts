import { useEffect, useMemo, useRef, useState } from 'react';

import type { SettingsHost, SettingValue } from '../types/settings';

export const useSetting = <T extends SettingValue = SettingValue>(
  host: SettingsHost | undefined,
  id: string,
) => {
  const hostRef = useRef(host);
  hostRef.current = host;

  const [currentValue, setCurrentValue] = useState<T | undefined>(undefined);

  useEffect(() => {
    if (!hostRef.current) return;
    let unsubscribe: (() => void) | undefined;
    let isMounted = true;

    hostRef.current.get<T>(id).then((initialValue) => {
      if (!isMounted) return;
      setCurrentValue(initialValue);
      unsubscribe = hostRef.current?.subscribe<T>(id, (nextValue) =>
        setCurrentValue(nextValue),
      );
    });

    return () => {
      isMounted = false;
      if (unsubscribe) unsubscribe();
    };
  }, [id, hostRef]);

  const setValue = useMemo(
    () => (nextValue: T) => {
      if (!hostRef.current) return;
      void hostRef.current.set<T>(id, nextValue);
    },
    [id],
  );

  return [currentValue, setValue] as const;
};
