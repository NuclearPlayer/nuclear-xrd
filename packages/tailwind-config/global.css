@import 'tailwindcss';
@import 'tw-animate-css';

@source "../player";
@source "../ui";
@source "../storybook";

@theme {
  --color-background: var(--background);
  --color-background-secondary: var(--background-secondary);
  --color-background-input: var(--background-input);

  --color-foreground: var(--foreground);
  --color-foreground-secondary: var(--foreground-secondary);
  --color-foreground-input: var(--foreground-input);

  --color-primary: var(--primary);
  --color-primary-hover: var(--primary-hover);
  --color-primary-active: var(--primary-active);

  --color-border: var(--border);
  --color-border-input: var(--border-input);
  --color-ring: var(--ring);

  --color-accent-green: var(--accent-green);
  --color-accent-yellow: var(--accent-yellow);
  --color-accent-purple: var(--accent-purple);
  --color-accent-blue: var(--accent-blue);
  --color-accent-orange: var(--accent-orange);
  --color-accent-cyan: var(--accent-cyan);
  --color-accent-red: var(--accent-red);

  --color-shadow: var(--shadow-color);

  --radius: var(--radius);
  --radius-sm: var(--radius-sm);
  --radius-lg: var(--radius-lg);

  --font-weight-normal: var(--font-weight-normal);
  --font-weight-bold: var(--font-weight-bold);

  --font-sans: var(--font-family);

  --shadow: var(--shadow-x) var(--shadow-y) 0px 0px var(--border);
  --shadow-shadow: var(--shadow);
  --shadow-nav: 4px 4px 0px 0px var(--border);
  --shadow-darkNav: 4px 4px 0px 0px var(--dark-border);
  --spacing-shadow-x: var(--shadow-x);
  --spacing-shadow-y: var(--shadow-y);
}

@custom-variant dark (&:where([data-theme=dark], [data-theme=dark] *));

:root {
  /* Core Backgrounds */
  --background: oklch(0.9491 0.023 341.75);
  --background-secondary: oklch(100% 0 0);
  --background-input: oklch(100% 0 0);

  /* Text Colors */
  --foreground: oklch(0% 0 0);
  --foreground-secondary: oklch(0.3 0 0);
  --foreground-input: oklch(0% 0 0);

  /* Interactive Elements */
  --primary: oklch(76.91% 0.173 341.75);
  --primary-hover: oklch(0% 0 0);
  --primary-active: oklch(0% 0 0);

  /* Borders & Outlines */
  --border: oklch(0% 0 0);
  --ring: oklch(100% 0 0);

  /* Accent Colors */
  --accent-green: oklch(79.05% 0.209 147.58);
  --accent-yellow: oklch(95.53% 0.134 112.76);
  --accent-purple: oklch(74.2% 0.149 301.88);
  --accent-blue: oklch(55.98% 0.08 270.09);
  --accent-orange: oklch(83.39% 0.124 66.56);
  --accent-cyan: oklch(88.26% 0.093 212.85);
  --accent-red: oklch(68.22% 0.206 24.43);

  /* Shadows */
  --shadow-color: oklch(0% 0 0);
  --shadow-x: 2px;
  --shadow-y: 2px;
  --shadow-blur: 0px;

  /* Typography */
  --font-family: 'DM Sans', system-ui, -apple-system, sans-serif;
  --font-weight-normal: 400;
  --font-weight-bold: 700;

  /* Border Radius */
  --radius: 8px;
  --radius-sm: 4px;
  --radius-lg: 12px;
}

@layer base {
  body {
    @apply text-foreground bg-background font-normal;
  }

  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    @apply font-bold;
  }
}

[data-theme='dark'] {
  --background: oklch(0.4408 0.0838 341.75);
  --background-secondary: oklch(39.49% 0.031 277.49);
  --background-input: oklch(0% 0 0);

  --foreground: oklch(0.9249 0 0);
  --foreground-secondary: oklch(0.9249 0 0);
  --foreground-input: oklch(100% 0 0);
}
