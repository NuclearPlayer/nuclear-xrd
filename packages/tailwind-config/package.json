{
  "name": "@nuclearplayer/tailwind-config",
  "version": "0.0.7",
  "description": "Shared Tailwind CSS configuration for Nuclear",
  "type": "module",
  "main": "./global.css",
  "exports": {
    ".": "./global.css"
  },
  "files": [
    "global.css"
  ],
  "scripts": {
    "lint": "eslint .",
    "lint:fix": "eslint . --fix"
  },
  "dependencies": {
    "@tailwindcss/vite": "^4.1.11",
    "tailwindcss": "^4.1.11",
    "tw-animate-css": "^1.3.6"
  },
  "devDependencies": {
    "@nuclearplayer/eslint-config": "workspace:*"
  }
}
