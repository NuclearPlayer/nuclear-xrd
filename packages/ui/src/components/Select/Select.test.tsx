import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it } from 'vitest';

import '@nuclearplayer/tailwind-config';

import { Select } from '.';

const OPTIONS = [
  { id: 'low', label: 'Low' },
  { id: 'medium', label: 'Medium' },
  { id: 'high', label: 'High' },
];

describe('Select', () => {
  it('(Snapshot) renders basic with description', () => {
    const { container } = render(
      <Select
        id="select-quality"
        label="Quality"
        options={OPTIONS}
        defaultValue="medium"
        description="Choose your preferred playback quality."
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('(Snapshot) renders error without description', () => {
    const { container } = render(
      <Select
        id="select-quality-error"
        label="Quality"
        options={OPTIONS}
        error="Please make a selection"
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('runs onValueChange callback on selection', async () => {
    const onValueChange = vi.fn();
    render(<Select options={OPTIONS} onValueChange={onValueChange} />);
    await userEvent.click(await screen.findByRole('button'));
    await userEvent.click(await screen.findByRole('option', { name: 'High' }));
    expect(onValueChange).toHaveBeenCalledWith('high');
  });
});
