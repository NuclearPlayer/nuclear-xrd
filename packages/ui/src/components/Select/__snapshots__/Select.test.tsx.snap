// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select snapshots > (Snapshot) renders basic with description 1`] = `
<div
  class="relative flex w-full flex-col gap-2"
>
  <label
    class="text-foreground text-sm font-semibold"
    for="select-quality"
    id="select-quality-label"
  >
    Quality
  </label>
  <button
    aria-controls="select-quality-listbox"
    aria-describedby="select-quality-description"
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="select-quality-label"
    class="border-border bg-primary text-foreground focus-visible:ring-primary relative flex h-11 w-full items-center justify-between rounded border-2 text-left transition-colors focus-visible:ring-2 focus-visible:outline-none px-3 py-2 text-sm"
    id="select-quality"
    type="button"
  >
    <span
      class="block truncate"
    >
      Medium
    </span>
    <span
      class="flex items-center"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-down text-foreground transition-transform"
        fill="none"
        height="16"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </span>
  </button>
  <p
    class="text-foreground-secondary text-sm select-none"
    id="select-quality-description"
  >
    Choose your preferred playback quality.
  </p>
</div>
`;

exports[`Select snapshots > (Snapshot) renders error without description 1`] = `
<div
  class="relative flex w-full flex-col gap-2"
>
  <label
    class="text-foreground text-sm font-semibold"
    for="select-quality-error"
    id="select-quality-error-label"
  >
    Quality
  </label>
  <button
    aria-controls="select-quality-error-listbox"
    aria-describedby="select-quality-error-error"
    aria-errormessage="select-quality-error-error"
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-invalid="true"
    aria-labelledby="select-quality-error-label"
    class="bg-primary text-foreground focus-visible:ring-primary relative flex h-11 w-full items-center justify-between rounded border-2 text-left transition-colors focus-visible:ring-2 focus-visible:outline-none px-3 py-2 text-sm border-accent-red"
    id="select-quality-error"
    type="button"
  >
    <span
      class="block truncate"
    >
      Low
    </span>
    <span
      class="flex items-center"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-down text-foreground transition-transform"
        fill="none"
        height="16"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </span>
  </button>
  <p
    class="text-accent-red text-xs select-none"
    id="select-quality-error-error"
  >
    Please make a selection
  </p>
</div>
`;
